### PR TITLE
os/bluestore: fix >2GB writes

### PR DESCRIPTION
**Pull request by liewegas**
This replaces #21326, which for some reason won't update when I force push.

- make KernelDevice break up aios that are larger than the max size
- add a unit test

### CODE REVIEW

**Comment by liewegas** [comment]
@kungf look ok?

File: src/test/objectstore/test_bluefs.cc [hunk 1]
```diff
@@ -135,6 +135,58 @@ TEST(BlueFS, small_appends) {
   rm_temp_bdev(fn);
 }
 
+TEST(BlueFS, very_large_write) {
```
Comment (ifed01) [inline src/test/objectstore/test_bluefs.cc]: This looks fine as a workaround to test the patch quickly.
But IMO there is a couple of general issues with this approach:
1) BlueFS involvement which makes large write occurrence implicit/unclear.
2) This doesn't cover other possible BlockDevice implementations: NVMEDevice/PMEMDevice and non-libaio mode. It's still unclear to me if the latter supports large writes.
I think we should consider making a standalone UT for BlockDevice implementations. May be later though...

File: src/test/objectstore/test_bluefs.cc [hunk 2]
```diff
@@ -135,6 +135,58 @@ TEST(BlueFS, small_appends) {
...
+    BlueFS::FileReader *h;
+    ASSERT_EQ(0, fs.open_for_read("dir", "bigfile", &h));
+    bufferlist bl;
+    BlueFS::FileReaderBuffer readbuf(10485760);
+    for (unsigned i = 0; i < 3*1024*1048576ull / sizeof(buf); ++i) {
+      bl.clear();
+      fs.read(h, &readbuf, i * sizeof(buf), sizeof(buf), &bl, NULL);
+      int r = memcmp(buf, bl.c_str(), sizeof(buf));
+      if (r) {
+	cout << "read got mismatch at offset " << i*sizeof(buf) << " r " << r
```
Comment (ifed01) [inline src/test/objectstore/test_bluefs.cc]: /s/cout/cerr?

**Comment by kungf** [comment]
hi @liewegas  does this pr need to backport to L ?

**Comment by liewegas** [comment]
@kungf N, M, and L, I think!

**Pull request merged**